### PR TITLE
WIP | =sink is now =move

### PR DESCRIPTION
**Pull request by Clyybber**
... which makes self-assignments work correctly.
Refs #11217

### CODE REVIEW

File: compiler/injectdestructors.nim [hunk 1]
```diff
@@ -410,8 +410,8 @@ proc destructiveMoveVar(n: PNode; c: var Con): PNode =
   add(v, vpart)
 
   result.add v
-  result.add genWasMoved(n, c)
   result.add tempAsNode
+  echo result
```
Comment (Araq) [inline compiler/injectdestructors.nim]: Debug code leftover.

File: compiler/liftdestructors.nim [hunk 2]
```diff
@@ -343,7 +343,7 @@ proc fillStrOp(c: var TLiftCtx; t: PType; body, x, y: PNode) =
 
 proc weakrefOp(c: var TLiftCtx; t: PType; body, x, y: PNode) =
   case c.kind
-  of attachedSink:
+  of attachedMove:
     # we 'nil' y out afterwards so we *need* to take over its reference
```
Comment (Araq) [inline compiler/liftdestructors.nim]: Well we don't anymore 'nil' y out, but we should.

File: compiler/liftdestructors.nim [hunk 3]
```diff
@@ -391,7 +391,7 @@ proc closureOp(c: var TLiftCtx; t: PType; body, x, y: PNode) =
     let xx = genBuiltin(c.graph, mAccessEnv, "accessEnv", x)
     xx.typ = getSysType(c.graph, c.info, tyPointer)
     case c.kind
-    of attachedSink:
+    of attachedMove:
       # we 'nil' y out afterwards so we *need* to take over its reference
```
Comment (Araq) [inline compiler/liftdestructors.nim]: Well we don't anymore 'nil' y out, but we should.

File: compiler/semstmts.nim [hunk 4]
```diff
@@ -1610,7 +1610,7 @@ proc semOverride(c: PContext, s: PSym, n: PNode) =
                  "signature for 'deepCopy' must be proc[T: ptr|ref](x: T): T")
     incl(s.flags, sfUsed)
     incl(s.flags, sfOverriden)
-  of "=", "=sink":
+  of "=", "=move":
```
Comment (Araq) [inline compiler/semstmts.nim]: but =move has the (var T, var T) signature and = doesn't.

File: lib/system.nim [hunk 5]
```diff
@@ -471,9 +471,10 @@ when defined(nimArrIdx):
   proc `=destroy`*[T](x: var T) {.inline, magic: "Destroy".} =
     ## Generic `destructor`:idx: implementation that can be overriden.
     discard
-  proc `=sink`*[T](x: var T; y: T) {.inline, magic: "Asgn".} =
+  proc `=move`*[T](x, y: var T) {.inline, magic: "Asgn".} =
     ## Generic `sink`:idx: implementation that can be overriden.
```
Comment (Araq) [inline lib/system.nim]: Please also change the doc comment.

**Comment by Araq** [comment]
Note: I'm taking over, I really want this in 0.20.

**Comment by Clyybber** [comment]
TODO:
* handle explicit moves better??
* figure out wether to elide the destructor for a `isAnalysableFieldAccess` or not

**Comment by Araq** [comment]
You need to patch the C codegen so that ``tySink`` can also be "pass by hidden pointer". Also this section

```nim

  if owner.kind in {skProc, skFunc, skMethod, skIterator, skConverter}:
    let params = owner.typ.n
    for i in 1 ..< params.len:
      let param = params[i].sym
      if isSinkTypeForParam(param.typ) and hasDestructor(param.typ.skipTypes({tySink})):
        c.destroys.add genDestroy(c, param.typ.skipTypes({tyGenericInst, tyAlias, tySink}), params[i])
``` 

needs to be removed, the caller is responsible for cleaning up what was passed as a sink parameter.

**Comment by Clyybber** [comment]
TODO:
* Make closures work
* Fix obj and tuple constructors of the backends: (affected test: tgcdestructors.nim)
  This is a bit tricky, see last commit for a dirty leaking unmergeable workaround.
  @Araq The problem is this:
```nim
type
  Obj* = object
    f*: seq[int]

var o = Obj(f: @[1, 2, 3])
```
  will be transformed to:
```c
colontmpD_ = TM_mDIVXahtYVExBX9bJLnfz0Q_76;
T2_.f = colontmpD_;
eqmove__C1DzWNOtypwtWid2xBvXLg((&o), (&T2_));
//...
eqdestroy__39bIYiz25XeqkqfTm4wBXNg((&colontmpD_));
eqdestroy__9aQE9amztaLIv7YEpj9agHaSg((&o));
```
  which will cause a double free segfault, since `colontmpD_` is shallow copied to the field of `T2_` / `o`.

**Comment by Clyybber** [comment]
This turned out to solve a problem that doesn't exist, and also hurt performance by 5% on avg.

**Pull request closed without merging**